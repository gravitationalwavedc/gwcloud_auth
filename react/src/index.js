import getRoutes from "./Routes";

export {
    getRoutes
};