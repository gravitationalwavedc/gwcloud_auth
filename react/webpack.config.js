const webpack = require("webpack");

module.exports = {
    module: {
        rules: [
            {
                test: /\.(js|jsx)$/,
                exclude: /node_modules/,
                use: {
                    loader: "babel-loader",
                    options: {
                        presets: [
                            "@babel/preset-env",
                            "@babel/preset-react"
                        ],
                        plugins: [
                            "@babel/plugin-proposal-class-properties"
                        ]
                    }
                }
            },
            {
                test: /\.html$/,
                use: [
                    {
                        loader: "html-loader"
                    }
                ]
            },
            {
                test: /\.scss$/,
                use: [
                    "style-loader",
                    "css-loader",
                    "sass-loader"
                ]
            },
            {
                test: /\.css$/,
                use: [
                    'style-loader',
                    'css-loader'
                ],
            },
            {
                test: /\.(png|jp(e*)g|svg)$/,
                use: [
                    {
                        loader: 'url-loader',
                        options: {
                            // limit: 8000, // Convert images < 8kb to base64 strings
                            name: 'images/[hash]-[name].[ext]'
                        }
                    }
                ]
            }

        ]
    },
    output: {
        publicPath: "/",
        globalObject: "this",
    },
    // Server Configuration options
    devServer: {
        host: '0.0.0.0',
        port: 8081,
        disableHostCheck: true,
        historyApiFallback: true,
        headers: {
            "Access-Control-Allow-Origin": "http://localhost:8080",
            "Access-Control-Allow-Credentials": "true",
            "Access-Control-Allow-Headers": "Content-Type, Authorization, x-id, Content-Length, X-Requested-With",
            "Access-Control-Allow-Methods": "GET, POST, PUT, DELETE, OPTIONS"
        }
    },
    plugins: [
        new webpack.DefinePlugin({
            'process.env': {
                NODE_ENV: JSON.stringify(process.env.NODE_ENV || 'development')
            }
        })
    ],
    resolve: {
        extensions: ['.js', '.jsx']
    },
};
