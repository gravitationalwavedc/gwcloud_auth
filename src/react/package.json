{
  "name": "gwcloud-auth",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "NODE_ENV=development webpack serve --mode development --hot",
    "start_win": "set \"NODE_ENV=development\" && webpack serve --mode development --hot ",
    "build": "NODE_ENV=production webpack --mode production",
    "build_win": "set \"NODE_ENV=production\" && webpack --mode production",
    "relay": "relay-compiler --src ./src --schema ./data/schema.json --extensions=js --extensions=jsx",
    "test": "jest",
    "test-ci": "jest --ci --reporters=default --reporters=jest-junit"
  },
  "keywords": [],
  "author": "",
  "license": "MIT",
  "dependencies": {
    "@babel/core": "^7.13.10",
    "@babel/plugin-proposal-class-properties": "^7.13.0",
    "@babel/preset-env": "^7.13.10",
    "@babel/preset-react": "^7.12.13",
    "@babel/runtime": "^7.13.10",
    "@jest/globals": "^26.6.2",
    "@popperjs/core": "^2.9.1",
    "@testing-library/react": "^11.2.5",
    "babel-jest": "^26.6.3",
    "babel-loader": "^8.2.2",
    "babel-plugin-relay": "^11.0.0",
    "babel-plugin-transform-class-properties": "^6.24.1",
    "bootstrap": "^4.6.0",
    "camelcase": "^6.2.0",
    "core-js": "^3.9.1",
    "css-loader": "^5.1.2",
    "eslint": "^7.22.0",
    "eslint-plugin-jest": "^24.3.1",
    "eslint-plugin-react": "^7.22.0",
    "formik": "^2.2.6",
    "found": "^0.6.0",
    "graphql": "^15.5.0",
    "identity-obj-proxy": "^3.0.0",
    "jest": "^26.6.3",
    "jest-junit": "^12.0.0",
    "jest-watch-typeahead": "^0.6.1",
    "jquery": "^3.6.0",
    "linq": "^3.2.3",
    "node-sass": "^5.0.0",
    "query-string": "^6.14.1",
    "react": "^17.0.1",
    "react-bootstrap": "^1.5.2",
    "react-dom": "^17.0.1",
    "react-google-recaptcha": "^2.1.0",
    "react-icons": "^4.2.0",
    "react-relay": "^11.0.0",
    "react-test-renderer": "^17.0.1",
    "relay-compiler": "^11.0.0",
    "relay-runtime": "^11.0.0",
    "relay-test-utils": "^11.0.0",
    "sass-loader": "^11.0.1",
    "style-loader": "^2.0.0",
    "url-loader": "^4.1.1",
    "webpack": "^5.25.1",
    "webpack-cli": "^4.5.0",
    "webpack-dev-server": "^3.11.2",
    "yup": "^0.32.9"
  },
  "jest": {
    "roots": [
      "<rootDir>/src"
    ],
    "collectCoverageFrom": [
      "src/**/*.{js,jsx,ts,tsx}",
      "!src/**/*.d.ts"
    ],
    "collectCoverage": true,
    "reporters": [
      "default",
      "jest-junit"
    ],
    "coverageReporters": [
      "text",
      "cobertura"
    ],
    "setupFilesAfterEnv": [
      "<rootDir>/src/setupTests.js"
    ],
    "testMatch": [
      "<rootDir>/src/**/__tests__/**/*.{js,jsx,ts,tsx}",
      "<rootDir>/src/**/*.{spec,test}.{js,jsx,ts,tsx}"
    ],
    "transform": {
      "^.+\\.(js|jsx|ts|tsx)$": "babel-jest",
      "^.+\\.css$": "<rootDir>/jest/cssTransform.js",
      "^(?!.*\\.(js|jsx|ts|tsx|css|json)$)": "<rootDir>/jest/fileTransform.js"
    },
    "transformIgnorePatterns": [
      "[/\\\\]node_modules[/\\\\].+\\.(js|jsx|ts|tsx)$",
      "^.+\\.module\\.(css|sass|scss)$"
    ],
    "modulePaths": [],
    "moduleNameMapper": {
      "^.+\\.module\\.(css|sass|scss)$": "identity-obj-proxy"
    },
    "moduleFileExtensions": [
      "web.js",
      "js",
      "web.ts",
      "ts",
      "web.tsx",
      "tsx",
      "json",
      "web.jsx",
      "jsx",
      "node"
    ],
    "watchPlugins": [
      "jest-watch-typeahead/filename",
      "jest-watch-typeahead/testname"
    ]
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^5.11.9"
  }
}
