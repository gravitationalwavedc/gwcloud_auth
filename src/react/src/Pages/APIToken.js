import React, { useState } from 'react';
import { commitMutation, createFragmentContainer, graphql } from 'react-relay';
import { harnessApi } from '../index';
import { CopyButton, SpoilerButton, CheckButton } from '../Components/CustomButtons';
import { Button, Grid, Container, Dropdown, DropdownButton, Row, Table } from 'react-bootstrap';
import { isEqual, without } from 'lodash';


const APIToken = (props) => {

    const allApps = ['Bilby']
    const [tokens, setTokens] = useState(props.data.apiTokens);

    const usedApps = tokens.map(token => token.app)
    const allAppsUsed = isEqual(allApps.sort(), usedApps.sort())

    const createToken = (app) => {
        commitMutation(harnessApi.getEnvironment('auth'), {
            mutation: graphql`mutation APITokenCreateMutation($input: CreateAPITokenInput!){
                createApiToken(input: $input) {
                  result {
                      app
                      token
                  }
                }
              }`,
            variables: {
                input: {
                    app: app
                }
            },
            onCompleted: (response, errors) => {
                if (errors) {
                    console.log(errors);
                } else {
                    setTokens(
                        [
                            ...tokens,
                            response.createApiToken.result
                        ]
                    );
                }
            },
        });
    };
    
    const revokeToken = (app) => {
        commitMutation(harnessApi.getEnvironment('auth'), {
            mutation: graphql`mutation APITokenRevokeMutation($input: RevokeAPITokenInput!){
                revokeApiToken(input: $input) {
                  result
                }
              }`,
            variables: {
                input: {
                    app: app,
                }
            },
            onCompleted: (response, errors) => {
                if (errors) {
                    console.log(errors);
                } else {
                    setTokens(
                        tokens.filter(token => token.app != app)
                    );
                }
            },
        });
    };

    return (
        <Container>
            <h1 className="pt-5 mb-4">
                API Tokens
                <span className='float-right'>
                    <DropdownButton
                        disabled={allAppsUsed}
                        title='Create Token'
                        variant='primary'
                    >
                        {
                            allApps.map((app, i) => (
                                <Dropdown.Item
                                    key={i}
                                    hidden={usedApps.includes(app)}
                                    onSelect={() => createToken(app)}
                                >
                                    {app}
                                </Dropdown.Item>
                            ))
                        }
                    </DropdownButton>
                </span>
            </h1>
            <Table as={Grid}>
                <thead>
                    <tr>
                        <th>App</th>
                        <th>Token</th>
                        <th className='text-center'>Action</th>
                    </tr>
                </thead>
                <tbody>
                    {
                        tokens.map((token, i) => (
                            <tr key={i}>
                                <td className='align-middle'>{token.app}</td>
                                <td className='align-middle'><SpoilerButton content='Click to show token...' hiddenContent={token.token}/></td>
                                <td className='align-middle text-center'>
                                    <CopyButton variant="text btn-link" content='Copy Token' copyContent={token.token}/>
                                    <CheckButton
                                        variant="text text-danger"
                                        content='Revoke Token'
                                        cancelContent='Revoke token?'
                                        onClick={() => revokeToken(token.app)}
                                    />
                                    {/* <Button variant="text text-danger" onClick={() => revokeToken(token.app)}>Revoke Token</Button> */}
                                </td>
                            </tr>
                        ))
                    }
                </tbody>
            </Table>
        </Container>
    );
};

export default createFragmentContainer(APIToken,
    {
        data: graphql`
            fragment APIToken_data on Query {
                apiTokens {
                    app
                    token
                }
            }
        `
    }
);
