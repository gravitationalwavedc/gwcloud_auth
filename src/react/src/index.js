import getRoutes from "./Routes";

let harnessApi = null;
let project = null;
let isGWLab = null;
let isGWLandscape = null;
let isGWCloud = null;

function setHarnessApi(api) {
    harnessApi = api;
    project = harnessApi.currentProject();
    isGWLab = project.name === 'GWLab';
    isGWLandscape = project.name === 'GWLandscape';
    isGWCloud = project.name === 'GWCloud';
}

export {
    getRoutes,
    setHarnessApi,
    harnessApi,
    isGWLab,
    isGWLandscape,
    isGWCloud
};
