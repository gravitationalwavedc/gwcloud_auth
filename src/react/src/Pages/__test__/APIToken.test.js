import { createMockEnvironment } from 'relay-test-utils';
import { setHarnessApi } from '../../index';
import React from 'react';
import { describe, expect, it, jest } from '@jest/globals';
import { QueryRenderer, graphql } from 'react-relay';
import { MockPayloadGenerator } from 'relay-test-utils';
import { render, waitFor } from '@testing-library/react';
import APIToken from '../APIToken';


describe('registerTests', () => {

    const TestRenderer = () => (
        <QueryRenderer
            environment={environment}
            query={graphql`
            query APITokenTestQuery($app: String!) @relay_test_operation {
              ...APIToken_data @arguments(app: $app)
            }
          `}
            variables={{
                app: "Bilby"
            }}
            render={({ error, props }) => {
                if (props) {
                    return <APIToken data={props} router={router}/>;
                } else if (error) {
                    return error.message;
                }
                return 'Loading...';
            }}
        />
    );

    const testToken = "TestTokenHasNoMeaning"

    const mockAPITokenReturn = {
        String() {
            return testToken
        },
    };

    it('API token page renders correctly', async () => {
        expect.hasAssertions();
        const { asFragment } = render(<TestRenderer />);
        await waitFor(() => environment.mock.resolveMostRecentOperation(operation =>
            MockPayloadGenerator.generate(operation, mockAPITokenReturn)
        ));
        expect(asFragment(<TestRenderer />)).toMatchSnapshot();
    });

    it('token is displayed in page', async () => {
        expect.hasAssertions();
        const { getByText } = render(<TestRenderer />);
        await waitFor(() => environment.mock.resolveMostRecentOperation(operation =>
            MockPayloadGenerator.generate(operation, mockAPITokenReturn)
        ));
        const token = getByText(testToken)
        expect(token).toBeInTheDocument();
    });
});




