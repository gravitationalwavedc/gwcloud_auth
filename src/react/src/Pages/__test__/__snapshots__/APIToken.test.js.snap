// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerTests API token page renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1
      class="pt-5 mb-4"
    >
      API Tokens
      <span
        class="float-right"
      >
        <div
          class="dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-primary"
            type="button"
          >
            Create Token
          </button>
        </div>
      </span>
    </h1>
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            App
          </th>
          <th>
            Token
          </th>
          <th
            class="text-center"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="align-middle"
          >
            TestTokenHasNoMeaning
          </td>
          <td
            class="align-middle"
          >
            <button
              class="fade show btn btn-text text-secondary"
              type="button"
            >
              Click to show token...
            </button>
          </td>
          <td
            class="align-middle text-center"
          >
            <button
              class="btn btn-text btn-link"
              type="button"
            >
              Copy Token
            </button>
            <button
              class="btn btn-text text-danger"
              type="button"
            >
              Revoke Token
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
